pymc_marketing/mmm/tvp.py

```diff
@@ -78,29 +164,54 @@ def time_varying_prior(
         Regression.
     """
 
-    if X_mid is None:
-        X_mid = float(X.mean().eval())
-    if L is None:
-        L = X_mid * 2
-
     model = pm.modelcontext(None)
 
-    if cov_func is None:
-        eta = pm.Exponential(f"{name}_eta", lam=eta_lam)
-        ls = pm.InverseGamma(f"{name}_ls", mu=ls_mu, sigma=ls_sigma)
-        cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)
+    eta = pm.Exponential(f"{name}_eta", lam=-np.log(eta_mass) / eta_upper)
+
+    if ls_upper is None:
+        ls = pc_prior_1d(f"{name}_ls", alpha=1.0 - ls_mass, lower=ls_lower)
+        ls_upper = 2 * X_mid
+    else:
+        # Note: this function sometimes fails, how to handle?
+        # If not using PC prior, should user set ls_mu and ls_sigma instead?
+        ls_params = pm.find_constrained_prior(
+            pm.InverseGamma,
+            lower=ls_lower,
+            upper=ls_upper,
+            mass=ls_mass,
+            init_guess={"alpha": 2, "beta": 1},
+        )
+        ls = pm.InverseGamma("f{name}_ls", **ls_params)
+
+    lengthscale_range = [ls_lower, ls_upper]
+    m, c = approx_hsgp_hyperparams(
+        X.eval(), X_mid, lengthscale_range, cov_func=cov_func
+    )
+    L = c * X_mid
 
-    model.add_coord("m", np.arange(m))  # type: ignore
+    model.add_coord("m", np.arange(m - 1))  # type: ignore
     hsgp_dims: str | tuple[str, str] = "m"
     if isinstance(dims, tuple):
         hsgp_dims = (dims[1], "m")
 
-    gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
+    cov_funcs = {
+        "expquad": pm.gp.cov.ExpQuad,
+        "matern52": pm.gp.cov.Matern52,
+        "matern32": pm.gp.cov.Matern32,
+    }
+    cov_func = eta**2 * cov_funcs[cov_func.lower()](input_dim=1, ls=ls)
+
+    gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func, drop_first=True)
     phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
-    hsgp_coefs = pm.Normal(f"{name}_hsgp_coefs", dims=hsgp_dims)
-    f = phi @ (hsgp_coefs * sqrt_psd).T
-    centered_f = f - f.mean(axis=0) + 1
```

Comment:
Do we want this to oscillate around `1.0` as before?

Reply:
I took out `f - f.mean(axis=0) + 1` since I think that will make predicting into the future difficult.  And a better way to zero mean this is to de-mean the basis `phi` directly.  I think if z ~ Normal(0, 1), then exp(z) will have a median at 1, so still oscillates around 1.0.